pymtl3/dsl/Component.py

```diff
@@ -55,10 +61,543 @@ def _construct( s ):
+                            lambda x: isinstance( x, Signal ), \
+                            lambda x: isinstance( x, MethodPort ) ] )
+
+      top._dsl.all_components    -= removed_components
+      top._dsl.all_signals       -= removed_signals
+      top._dsl.all_method_ports  -= removed_method_ports
+
+      top._dsl.all_named_objects -= removed_components
+
+      removed_connectables = removed_signals | removed_method_ports
+      top._dsl.all_named_objects -= removed_connectables
+
+      if isinstance( foo, Placeholder ):
+        # No need to uncollect vars from a placeholder
+        assert not foo._dsl.consts
+      else:
+        for x in removed_components:
+          # remove consts
+          for y in x._dsl.consts:
+            del y._dsl.parent_obj
+          # uncollect variables
+          top._uncollect_vars( x )
+
+      for x in removed_components:
+        del x._dsl.parent_obj
+      for x in removed_connectables:
+        del x._dsl.parent_obj
+
+      saved_connections = []
+
+      for x in removed_connectables:
+        # Clean up all_adjancency at top
+        if x in top._dsl.all_adjacency:
+          for other in top._dsl.all_adjacency[x]:
+            # other must be in the dict
+            # If other will be removed, we don't need to remove it here ..
+            if other not in removed_connectables:
+              top._dsl.all_adjacency[other].remove( x )
+              saved_connections.append( (other, "top"+repr(x)[1:]) ) # other is from outside
+          del top._dsl.all_adjacency[x]
+
+        # Clean up adjacency at parent
+        if x in parent._dsl.adjacency:
+          for other in parent._dsl.adjacency[x]:
+            # other must be in the dict
+            if other not in removed_connectables:
+              parent._dsl.adjacency[other].remove( x )
+          del parent._dsl.adjacency[x]
+
+      # We don't break nets anymore. Instead, we set the flags to true so
+      # that the next get_xxx_net will immediately recollect nets.
+      top._dsl._has_pending_value_connections = True
+      top._dsl._has_pending_method_connections = True
+
+      # We clean up the connect_order list. If we want to preserve the
+      # original connect order, we can play some other tricks here such as
+      # leaving the strings in there. However I don't think that's clean
+      # enough. Thus I'm just removing them right now.
+      new_connect_order = []
+      for (x, y) in parent._dsl.connect_order:
+        if x not in removed_signals and y not in removed_signals: # TODO method port
+          new_connect_order.append( (x, y) )
+
+      parent._dsl.connect_order = new_connect_order
+
+      return saved_connections
+
+    return _delete_component_internal( top, obj )
+    # import gc
+    # gc.collect() # this takes 0.1 seconds
+
+  #-----------------------------------------------------------------------
+  # Post-elaborate public APIs (can only be called after elaboration)
```

Comment:
I like the new API organization so much -- you don't have to dig into different levels to find out them!